### ISSUE TITLE
Incorrect type for addHook callback currentNode argument

### ISSUE THREAD

**Issue by benjdevries** [cure53/DOMPurify#1029]
> This issue proposes a [bug, feature] which... Bug

### Background & Context

`addHook` callback now typed with `Node` instead of `Element`

### Bug

In the latest version (3.0.5) of the DefinitelyTyped types for the `addHook` method, the `currentNode` argument to the callback is typed as `Element`.

```ts
// @types/dompurify/index.d.ts

addHook(hook: HookName, cb: (currentNode: Element, data: HookEvent, config: Config) => void): void;
```

In the first-party types bundled with the 3.2.0 release, this changes to `Node`

```ts
// DOMPurify/dist/purify.cjs.d.ts
...
addHook(entryPoint: BasicHookName, hookFunction: Hook): void;
...
type Hook = (this: DOMPurify, currentNode: Node, hookEvent: null, config: Config) => void;
```
 
Its possible this was intentional since `Element` extends `Node` and may not be appropriate for all hooks; however, in doing so, you have broken one of your own examples.  https://github.com/cure53/DOMPurify/blob/f93acffc37667f7e6a008d7396c961beba831c92/demos/hooks-target-blank-demo.html

```ts
// Add a hook to make all links open a new window
DOMPurify.addHook('afterSanitizeAttributes', (node) => {  // node is now typed as Node, not Element
    if ('target' in node) {
        node.setAttribute('target', '_blank');  // property setAttribute does not exist on type Node
        node.setAttribute('rel', 'noopener noreferrer');
    }
    if (!node.hasAttribute('target') && (node.hasAttribute('xlink:href') || node.hasAttribute('href'))) {  // property hasAttribute does not exist on type Node
       node.setAttribute('xlink:show', 'new');
    }
});
```

`setAttribute` and `getAttribute` are implemented on `Element`, not `Node`, so this example no longer satisfies TypeScript.

Would like some clarification on this. Was this change intentional? If so, that means that hooks must be written to support general `Node` objects and not just `Element` objects. Do you have an updated example of setting the target/rel attributes in a link element?

**Comment by cure53** [comment]
cc @reduckted please, who worked on the types

**Comment by cure53** [comment]
Thanks a lot, @reduckted .

@benjdevries Maybe it makes sense if we add an example file that showcases this situation exactly? I am hesitant to modify the existing ones but happy to add one to clarify.

**Comment by reduckted** [comment]
Just a note about my example. I picked `uponSanitizeElement` as the hook to demonstrate, but most of the other hooks have the same behavior  - they will be called with any type of node. 

`afterSanitizeAttributes`, however, does not behave that way, not does `uponSanitizeAttribute`. They will only be called if the node has an `attributes` property (something that only an `Element` has):
https://github.com/cure53/DOMPurify/blob/8e1c70a083183483eca5b8786f1217808856b394/src/purify.ts#L1285-L1294

Perhaps those two hooks could be changed to use an `Element` instead of a `Node`?

**Comment by cure53** [comment]
Do we need to change anything on our end or all fine? :sweat_smile:

**Comment by benjdevries** [comment]
For me personally, I'm fine with using a type guard since the change from `Element` to `Node` was intentional. Although it sounds like certain hooks will only be called on `Element` objects, so it might be good to restrict the typing on those so that a type guard isn't necessary.

As for the "broken" example that I pointed out (setting links to open in new tabs), it sounds like this is still a valid example since it will called on only nodes with attributes, so `hasAttribute` and `setAttribute` should be safe. It's just that when used with TypeScript, that example will need the type guard in order to pass the type checker. I'm not sure if any other examples are like this, I just picked that one since it was very similar to an usage in one of my projects.
